indispensable-josef/src/commonMain/kotlin/at/asitplus/signum/indispensable/josef/JweEncryption.kt

```diff
@@ -15,43 +18,51 @@ import kotlinx.serialization.encoding.Encoder
  * and also [RFC 7518](https://datatracker.ietf.org/doc/html/rfc7518)
  */
 @Serializable(with = JweEncryptionSerializer::class)
-enum class JweEncryption(val text: String) {
-
-    A128GCM("A128GCM"),
-    A192GCM("A192GCM"),
-    A256GCM("A256GCM"),
-    A128CBC_HS256("A128CBC-HS256"),
-    A192CBC_HS384("A192CBC-HS384"),
-    A256CBC_HS512("A256CBC-HS512")
+enum class JweEncryption(val identifier: String, val algorithm: SymmetricEncryptionAlgorithm<*, *, *>) {
+
+    A128GCM("A128GCM", SymmetricEncryptionAlgorithm.AES_128.GCM),
+    A192GCM("A192GCM", SymmetricEncryptionAlgorithm.AES_192.GCM),
+    A256GCM("A256GCM", SymmetricEncryptionAlgorithm.AES_256.GCM),
+    A128CBC_HS256("A128CBC-HS256", SymmetricEncryptionAlgorithm.AES_128.CBC.HMAC.SHA_256),
+    A192CBC_HS384("A192CBC-HS384", SymmetricEncryptionAlgorithm.AES_192.CBC.HMAC.SHA_384),
+    A256CBC_HS512("A256CBC-HS512", SymmetricEncryptionAlgorithm.AES_256.CBC.HMAC.SHA_512)
     ;
 
-    val encryptionKeyLength
-        get() = when (this) {
-            A128GCM -> 128
-            A192GCM -> 192
-            A256GCM -> 256
-            A128CBC_HS256 -> 256
-            A192CBC_HS384 -> 384
-            A256CBC_HS512 -> 512
-        }
 
-    val ivLengthBits
-        get() = when (this) {
-            A128GCM, A192GCM, A256GCM -> 96 // GCM: 96 bits
-            A128CBC_HS256, A192CBC_HS384, A256CBC_HS512 -> 128 // all AES-based
-        }
+    @Deprecated("Clumsy name", ReplaceWith("identifier"))
+    val text get() = identifier
 
     /**
-     * Per [RFC 7518](https://datatracker.ietf.org/doc/html/rfc7518#section-5.2.3),
-     * where the MAC output bytes need to be truncated to this size for use in JWE.
+     * For integrated AEAD algorithms, this is the length of the sole key.
+     * For bolted-on AEAD algorithms with a dedicated MAC key, such as AES-CBC+HMAC,
+     * this is the **length of the encryption key without the dedicated MAC key**.
      */
-    val macLength: Int?
-        get() = when (this) {
-            A128CBC_HS256 -> 16
-            A192CBC_HS384 -> 24
-            A256CBC_HS512 -> 32
-            else -> null
+    val encryptionKeyLength: BitLength get() = algorithm.keySize
+
+    val ivLength: BitLength
+        get() = when (algorithm.requiresNonce()) {
+            true -> algorithm.nonceTrait.length
+            false -> 0.bit
         }
+
+    /**
+     * for integrated AEAD algorithms, this is zero.
+     * For bolted-on AEAD algorithms with a dedicated MAC, this behaves as the name implies
+     */
+    val dedicatedMacKeyLength: BitLength get() = if (algorithm.hasDedicatedMac()) algorithm.preferredMacKeyLength else 0.bit
```

Comment:
this (and other stuff near it) should be extensions on the `Specialized...` marker interface

Reply:
not sure

Reply:
#249